cpp/include/cudf/io/orc_metadata.hpp

```diff
@@ -38,18 +51,130 @@ namespace io {
+  auto *get_sum() const { return sum.get(); }
+};
+
+struct integer_statistics : minmax_statistics<int64_t>, sum_statistics<int64_t> {
+};
+
+struct double_statistics : minmax_statistics<double>, sum_statistics<double> {
+};
+
+// According to ORC specs, the sum should be signed, but pyarrow uses unsigned value
+struct string_statistics : minmax_statistics<std::string>, sum_statistics<uint64_t> {
+};
+
+struct bucket_statistics {
+  std::vector<uint64_t> count;
+
+  auto *get_count(size_t index) const { return &count.at(index); }
+};
+
+struct decimal_statistics : minmax_statistics<std::string>, sum_statistics<std::string> {
+};
+
+struct date_statistics : minmax_statistics<int32_t> {
+};
+
+struct binary_statistics : sum_statistics<int64_t> {
+};
+
+struct timestamp_statistics : minmax_statistics<int64_t> {
+  std::unique_ptr<int64_t> minimumUtc;
+  std::unique_ptr<int64_t> maximumUtc;
+
+  auto has_minimumUtc() const { return minimumUtc != nullptr; }
+  auto has_maximumUtc() const { return maximumUtc != nullptr; }
+  auto *get_minimumUtc() const { return minimumUtc.get(); }
+  auto *get_maximumUtc() const { return maximumUtc.get(); }
+};
+
+/**
+ * @brief Enumerator for types of column statistics that can be included in `columnStatistics`.
+ *
+ * Different statistics types are generated for different column data types.
  */
-std::vector<std::vector<std::string>> read_orc_statistics(source_info const& src_info);
+enum class statistics_type {
+  NONE,
+  INT,
+  DOUBLE,
+  STRING,
+  BUCKET,
+  DECIMAL,
+  DATE,
+  BINARY,
+  TIMESTAMP,
+};
+
+/**
+ * @brief Contains per-column ORC statistics.
+ *
+ * `std::unique_ptr` is used to wrap the optional values.
+ * At most one of the ***_statistics members has a non-null value. The `type` member can se used to
+ * find the valid more easily.
+ */
+struct column_statistics {
+  std::unique_ptr<uint64_t> numberOfValues;
+  statistics_type type = statistics_type::NONE;
+  std::unique_ptr<integer_statistics> intStatistics;
+  std::unique_ptr<double_statistics> doubleStatistics;
+  std::unique_ptr<string_statistics> stringStatistics;
+  std::unique_ptr<bucket_statistics> bucketStatistics;
+  std::unique_ptr<decimal_statistics> decimalStatistics;
+  std::unique_ptr<date_statistics> dateStatistics;
+  std::unique_ptr<binary_statistics> binaryStatistics;
+  std::unique_ptr<timestamp_statistics> timestampStatistics;
+  // TODO: hasNull (issue #7087)
+};
+
+/**
+ * @brief Holds column names and parsed file-level and stripe-level statistics.
+ *
+ * The `column_names` and `column_stats` members contain one element per column. The `stripe_stats`
+ * contains one element per stripe, where each element contains column statistics for each column.
+ */
+struct parsed_orc_statistics {
+  std::vector<std::string> column_names;
+  std::vector<column_statistics> column_stats;
+  std::vector<std::vector<column_statistics>> stripe_stats;
+};
+
+parsed_orc_statistics read_parsed_orc_statistics(source_info const &src_info);
```

Comment:
Might be better to name this one simply `read_orc_statistics`, as it would match the Python API name. 
I went with this to be more explicit, but I'm definitely open to suggestions.